### PR TITLE
[AArch64][GlobalISel] Vector Constant Materialization

### PR DESCRIPTION
**Pull request by MarkAHarley**
Vector constants are always lowered via constant pool loads. This patch selects MOVI/MVNI in more cases where appropriate.

### CODE REVIEW

**Review by aemerson — APPROVED** [review]
Thanks, LGTM. I ran the CTMark suite at -Os and found these improvements:
```
Program                                       size.__TEXT,__text
                                              outputf8k03nj5     outputncrrixth diff
kimwitu++/kc                                  450660.00          450660.00       0.0%
SPASS/SPASS                                   410592.00          410588.00      -0.0%
mafft/pairlocalalign                          243700.00          243680.00      -0.0%
sqlite3/sqlite3                               287312.00          287288.00      -0.0%
consumer-typeset/consumer-typeset             412128.00          412084.00      -0.0%
tramp3d-v4/tramp3d-v4                         364108.00          364064.00      -0.0%
Bullet/bullet                                 450772.00          450688.00      -0.0%
lencod/lencod                                 427716.00          427620.00      -0.0%
ClamAV/clamscan                               381428.00          381340.00      -0.0%
7zip/7zip-benchmark                           566508.00          566236.00      -0.0%
                           Geomean difference                                   -0.0%
```

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 1]
```diff
@@ -5433,6 +5448,188 @@ bool AArch64InstructionSelector::selectInsertElt(MachineInstr &I,
...
+    auto Mov = builder.buildInstr(Op, {Dst}, {}).addImm(val);
+    constrainSelectedInstRegOperands(*Mov, TII, TRI, RBI);
+    return &*Mov;
+  }
+  return nullptr;
+}
+
+MachineInstr *AArch64InstructionSelector::tryAdvSIMDModImm16(
+    Register Dst, unsigned DstSize, APInt Bits, MachineIRBuilder &builder,
+    bool inv) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5475]: Can you change builder to Builder and inv to Inv. The llvm coding style uses capitalized variable names.
Same for the other variables like val below

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 2]
```diff
@@ -5433,6 +5448,188 @@ bool AArch64InstructionSelector::selectInsertElt(MachineInstr &I,
...
+    Shift = 8;
+  } else if ((isAdvSIMDModImm = AArch64_AM::isAdvSIMDModImmType3(val))) {
+    val = AArch64_AM::encodeAdvSIMDModImmType3(val);
+    Shift = 16;
+  } else if ((isAdvSIMDModImm = AArch64_AM::isAdvSIMDModImmType4(val))) {
+    val = AArch64_AM::encodeAdvSIMDModImmType4(val);
+    Shift = 24;
+  }
+
+  if (isAdvSIMDModImm) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5538]: I think I would personally do:
```
if (AArch64_AM::isAdvSIMDModImmType1(val)) {
 ..
} else if (..) {
 //
} else
  return nullptr;
```
That way after the if's we know we are creating the instruction, and the isAdvSIMDModImm hopefully isn't needed any more.

**Review by davemgreen — COMMENTED** [review]
> I ran the CTMark suite at -Os and found these improvements:

That's good to hear, you wouldn't expect a lot of vectors at Os.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 2 above)
Comment (MarkAHarley) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: Fixed

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 1 above)
Comment (MarkAHarley) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: Fixed

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Comment by davemgreen** [comment]
Committed in eb96d6e2fb570e153186b08e4148e8d1b07681e3.

**Pull request closed without merging**